### ISSUE TITLE
Create a code generation option for the method readers

### ISSUE THREAD

**Issue by RobAustin** [OpenHFT/Chronicle-Wire#202]
For now, method reader parses event name, creates instances of object arguments (they are created under the hood via ValueIn#object) and executes expected method via reflection (VanillaMethodReader#invoke).
The idea is to generate code of MethodReader dynamically instead of using VanillaMethodReader. Execution of generated code is expected to be more efficient as:

1. Instead of Method#invoke, generated code may execute regular Java method call as per we know argument types at the moment of MethodReader generation.
2. Instead of creating instances for object arguments, we can allocate objects once for every created MethodReader. Generated code sample below reuses a passed object during deserialization.
```
    // one for every possible method with a Marshallable
    SomeDTO someDTO = new SomeDTO();
    private static boolean methodReader0(final ExcerptTailer tailer, Printer printer) {
        try (DocumentContext dc = tailer.readingDocument()) {
            if (!dc.isPresent())
                 return false;

            String eventName = dc.wire().readEventName(String.class);
            switch(eventName) {
                case "print":
    // no marshallable to reuse
                    String arg = dc.wire().getValueIn().text();
                    printer.print(arg);
                    break;
                case "someDTO":
    // read using the reusable marshallable
                    SomeDTO someDTO = dc.wire().getValueIn().object(this.someDTO);
                    printer.someDTO(someDTO);
                    break;
               default:
                    // optional warning of unexpected methodName

        }
    }
```
Generated code should also support:
1. Method chaining (see ChainedMethodsTest for the details)
2. Effective long serialization for methods with one long parameter
3. Method reader interceptor invocation, if present. Interceptor will still invoke methods via reflection. **Note: code generation for MethodWriters doesn't support interceptors so far, so I guess this is optional.**
4. MethodFilterOnFirstArg logic
5. Efficient method name serialization with @MethodId

**Comment by glukos** [comment]
For now, method reader parses event name, creates instances of object arguments (they are created under the hood via ValueIn#object) and executes expected method via reflection (VanillaMethodReader#invoke).
The idea is to generate code of MethodReader dynamically instead of using VanillaMethodReader. Execution of generated code is expected to be more efficient as:
1. Instead of Method#invoke, generated code may execute regular Java method call as per we know argument types at the moment of MethodReader generation.
2. Instead of creating instances for object arguments, we can allocate objects once for every created MethodReader. Generated code sample below reuses a passed object during deserialization.
`
// one for every possible method with a Marshallable
         SomeDTO someDTO = new SomeDTO();
    private static boolean methodReader0(final ExcerptTailer tailer, Printer printer) {
        try (DocumentContext dc = tailer.readingDocument()) {
            if (!dc.isPresent())
                 return false;

            String eventName = dc.wire().readEventName(String.class);
            switch(eventName) {
                case "print":
// no marshallable to reuse
                    String arg = dc.wire().getValueIn().text();
                    printer.print(arg);
                    break;
                case "someDTO":
// read using the reusable marshallable
                    SomeDTO someDTO = dc.wire().getValueIn().object(this.someDTO);
                    printer.someDTO(someDTO);
                    break;
               default:
                    // optional warning of unexpected methodName

        }
    }
`
Generated code should also support:
1. Method chaining (see ChainedMethodsTest for the details)
2. Effective long serialization for methods with one long parameter
3. Method reader interceptor invocation, if present (interceptor will still invoke methods via reflection)
4. MethodFilterOnFirstArg

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.20.66](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.20.66), [BOM-2.20.86](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.86)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.20.101](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.20.101), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
